code/controllers/subsystem/polling.dm

```diff
@@ -0,0 +1,312 @@
+	var/time_started // The world.time at which the poll was created
+	var/finished = FALSE // Whether the polling is finished
+	var/hash // Used to categorize in the alerts system
+
+/datum/candidate_poll/New(polled_role, polled_question, poll_duration, poll_ignoring_category, poll_jumpable)
+	role = polled_role
+	question = polled_question
+	duration = poll_duration
+	ignoring_category = poll_ignoring_category
+	jump_to_me = poll_jumpable
+	signed_up = list()
+	time_started = world.time
+	hash = copytext(md5("[question]_[role ? role : "0"]"), 1, 7)
+	return ..()
+
+/datum/candidate_poll/proc/sign_up(mob/M, silent = FALSE)
+	if(!istype(M) || !M.key || !M.client)
+		return FALSE
+	if(M in signed_up)
+		if(!silent)
+			to_chat(M, span_warning("You have already signed up for this!"))
+		return FALSE
+	if(time_left() <= 0)
+		if(!silent)
+			to_chat(M, span_danger("Sorry, you were too late for the consideration!"))
+			SEND_SOUND(M, 'sound/machines/buzz-sigh.ogg')
+		return FALSE
+
+	signed_up += M
+	if(!silent)
+		to_chat(M, span_notice("You have signed up for [role]! A candidate will be picked randomly soon."))
+		// Sign them up for any other polls with the same mob type
+		for(var/existing_poll in SSpolling.currently_polling)
+			var/datum/candidate_poll/P = existing_poll
+			if(src != P && hash == P.hash && !(M in P.signed_up))
+				P.sign_up(M, TRUE)
+	if(alert_button)
+		alert_button.update_candidates_number_overlay()
+	return TRUE
+
+/datum/candidate_poll/proc/remove_candidate(mob/M, silent = FALSE)
+	if(!istype(M) || !M.key || !M.client)
+		return FALSE
+	if(!(M in signed_up))
+		if(!silent)
+			to_chat(M, span_warning("You aren't signed up for this!"))
+		return FALSE
+
+	if(time_left() <= 0)
+		if(!silent)
+			to_chat(M, span_danger("It's too late to unregister yourself, selection has already begun!"))
+		return FALSE
+
+	signed_up -= M
+	if(!silent)
+		to_chat(M, span_danger("You have been unregistered as a candidate for [role]. You can sign up again before the poll ends."))
+
+		for(var/existing_poll in SSpolling.currently_polling)
+			var/datum/candidate_poll/P = existing_poll
+			if(src != P && hash == P.hash && (M in P.signed_up))
+				P.remove_candidate(M, TRUE)
+	if(alert_button)
+		alert_button.update_candidates_number_overlay()
+	return TRUE
+
+/datum/candidate_poll/proc/never_for_this_round(mob/M, undoing = FALSE)
+	if(!undoing)
+		var/list/ignore_list = GLOB.poll_ignore[ignoring_category]
+		if(!ignore_list)
+			GLOB.poll_ignore[ignoring_category] = list()
+		GLOB.poll_ignore[ignoring_category] += M.ckey
+		to_chat(M, span_danger("Choice registered: Never for this round."))
+		remove_candidate(M, silent = TRUE)
+		return
+	GLOB.poll_ignore[ignoring_category] -= M.ckey
+	to_chat(M, span_notice("Choice registered: Eligible for this round"))
+
+/datum/candidate_poll/proc/trim_candidates()
+	list_clear_nulls(signed_up)
+	for(var/mob in signed_up)
+		var/mob/M = mob
+		if(!M.key || !M.client)
+			signed_up -= M
```

Comment:
```suggestion
	for(var/mob/player in signed_up) // are we assuming that `signed_up` has nulls in the list? if not, add `as anything`
		if(isnull(player.key) || isnull(player.client))
			signed_up -= player
```

A) we don't need to typecast outside the for loop.
B) as anything saves any perf concerns (unless we might expect to get nulls or something in this list)
C) `isnull()` is more explicit when it comes to typed references like `client` or stuff that defaults to null like `key`. it's also marginally faster.